src/libdredd/src/mutate_visitor.cc

```diff
@@ -108,14 +108,23 @@ bool MutateVisitor::TraverseDecl(clang::Decl* decl) {
     // consider it for mutation.
     return true;
   }
-  if (llvm::dyn_cast<clang::StaticAssertDecl>(decl) != nullptr) {
+  if (const auto* static_assert_decl =
+          llvm::dyn_cast<clang::StaticAssertDecl>(decl)) {
     // It does not make sense to mutate static assertions, as (a) this will
     // very likely lead to compile-time failures due to the assertion not
     // holding, (b) if compilation succeeds then the assertion is not actually
     // present at runtime so there is no notion of killing the mutant, and (c)
     // dynamic instrumentation of the mutation operator will break the rules
     // associated with static assertions anyway.
-    return true;
+
+    // However, some C++ compiler complain if the expression under static
+    // assertion is not a compile-time constants. To allow the constant
+    // declaration that occurs inside static assertion to have their initial
+    // value mutated (because they may be used elsewhere), but to avoid
+    // compilaton errors, we record the static assertion so that later their
+    // argument expression can be replaced with the value to which the argument
+    // expression would notmally evaluate.
+    static_assertions_to_rewrite_.push_back(static_assert_decl);
```

Comment:
Before the change there was `return true`, which ensures that the expression associated with the static cast won't be recursed into.

You have replaced this with the logging that this rewrite is needed - but shouldn't both happen - i.e., logging, and then still doing a return? I think that with your current change the expression of the static assertion will be mutated, but then over-written. Could you investigate?

Reply:
You are right that not having return here caused the expression under it to be recursed into. However, these expressions are not mutated because `MutateVisitor::VisitExpr` returns early due to these expressions not being `IsInFunction()`.

Now, I am not sure why `MutateVisitor::IsInFunction()` considers these expressions not inside a function. This could point to a bug in `MutateVisitor::IsInFunction()`.

Reply:
The AST is as below:
```
-FunctionDecl 0x555558e59920 </home/ubuntu/dredd/examples/simple/assert.cc:1:1, line:3:1> line:1:5 main 'int ()'
  `-CompoundStmt 0x555558e59af0 <col:12, line:3:1>
    `-DeclStmt 0x555558e59ad8 <line:2:3, col:19>
      `-StaticAssertDecl 0x555558e59a98 <col:3, col:18> col:3
        |-ImplicitCastExpr 0x555558e59a60 <col:17> 'bool' <IntegralToBoolean>
        | `-IntegerLiteral 0x555558e59a40 <col:17> 'int' 1
        `-<<<NULL>>>
```
Note that static assertion is a declaration. 

If the program doesn't return at that point, the it will proceed with adding `StaticAssertDecl` to global variable `enclosing_decls_`. 

When the integer literal is traversed by  `MutateVisitor::VisitExpr`, it will check whether this expression is in a function(`MutateVisitor::IsInFunction()`). `MutateVisitor::IsInFunction()` loop through each enclosing declaration and check
(1) if the enclosing decl is a function declaration, return true
(2) if it is not a variable declaration, return false
(3) otherwise, continue looping 
Thus, the integer literal is "not in function" because the enclosing scope - `StaticAssertDecl` - is not a variable declaration.

Reply:
Interesting. I wonder then why I had the previous "return true" - it seems like I wouldn't have even needed the check for a static assertion given what you have discovered about `IsInFunction`. But also it seems unlikely that I would have added this for no reason.

Reply:
Seems like there's no `MutateVisitor::IsInFunction()` back when the check for a static assertion is first written. The enclosing declaration is used in a simpler way. 
```
if (enclosing_decls_.empty()) {
    // Only consider mutating binary expressions that occur inside functions.
    return true;
  }
```

Reply:
Thanks for investigating the revision history here - and what you have done looks good.